commands/command_filter_process.go

```diff
@@ -116,6 +212,135 @@ func filterCommand(cmd *cobra.Command, args []string) {
+// if available had an infinite channel buffer.
+func infiniteTransferBuffer(q *tq.TransferQueue, available chan<- *tq.Transfer) {
+	// Stream results from q.Watch() into chan "available" via an infinite
+	// buffer.
+
+	watch := q.Watch()
+
+	// pending is used to keep track of an ordered list of available
+	// `*tq.Transfer`'s that cannot be written to "available" without
+	// blocking.
+	var pending []*tq.Transfer
+
+	for {
+		if len(pending) > 0 {
+			select {
+			case t, ok := <-watch:
+				if !ok {
+					// If the list of pending elements is
+					// non-empty, stream them out (even if
+					// they block), and then close().
+					for _, t = range pending {
+						available <- t
+					}
+					close(available)
+					return
+				}
+				pending = append(pending, t)
+			case available <- pending[0]:
+				// Otherwise, dequeue and shift the first
+				// element from pending onto available.
+				pending = pending[1:]
+			}
+		} else {
+			t, ok := <-watch
+			if !ok {
+				// If watch is closed, the "tq" is done, and
+				// there are no items on the buffer.  Return
+				// immediately.
+				close(available)
+				return
+			}
+
+			select {
+			case available <- t:
+			// Copy an item directly from <-watch onto available<-.
+			default:
+				// Otherwise, if that would have blocked, make
+				// the new read pending.
+				pending = append(pending, t)
+			}
+		}
+	}
+}
+
+// incomingOrCached returns an io.Reader that is either the contents of the
+// given io.Reader "r", or the encoded contents of "ptr". It returns an error if
+// there was an error reading from "r".
+//
+// This is done because when a `command=smudge` with `can-delay=0` is issued,
+// the entry's contents are not sent, and must be re-encoded from the stored
+// pointer corresponding to the request's filepath.
+func incomingOrCached(r io.Reader, ptr *lfs.Pointer) (io.Reader, error) {
+	buf := make([]byte, 1024)
+	n, err := r.Read(buf)
+	buf = buf[:n]
+
+	if n == 0 {
+		if ptr == nil {
+			// If we read no data from the given io.Reader "r" _and_
+			// there was no data to fall back on, return an empty
+			// io.Reader yielding no data.
+			return bytes.NewReader(buf), nil
+		}
+		// If we read no data from the given io.Reader "r", _and_ there
+		// is a pointer that we can fall back on, return an io.Reader
+		// that yields the encoded version of the given pointer.
+		return strings.NewReader(ptr.Encoded()), nil
+	}
+
+	if err == io.EOF {
+		return bytes.NewReader(buf), nil
+	}
+	return io.MultiReader(bytes.NewReader(buf), r), err
+}
+
+// readAvailable satisfies the accumulation semantics for the
+// 'list_available_blobs' command. It accumulates items until:
+//
+// 1. Reading from the channel of available items blocks, or ...
+// 2. There is one item available, or ...
+// 3. The 'tq.TransferQueue' is completed.
+func readAvailable(ch <-chan *tq.Transfer) []*tq.Transfer {
+	ts := make([]*tq.Transfer, 0, 100)
```

Comment:
Since the transfer queue has a batch size of 100, and `Watch()` is buffered at 100 items, all 100 could be ready at once. This is done to prevent growing the array.